pkg/reconciler/v1alpha1/revision/resources/env_var.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"github.com/knative/serving/pkg/logging"
+	"github.com/knative/serving/pkg/reconciler/v1alpha1/revision/resources/names"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	knativeEnvVariableKey = "KNATIVE_CONTEXT"
```

Comment:
Why do you prefer one env var as opposed to multiple env vars (as suggested in #1787)
From a developer perspective, I think I am more in favor being able directly to get the Revision name from an env var, instead of having to parse JSON first.

Reply:
It was suggested in the issue comments to use a similar pattern to `cloud foundry` and `spring`, but I also feel it has the advantage of having system generated env variables contained in one spot which also makes it easier to expand in the future. From a developer perspective, I found this pattern useful when using it before.

Reply:
I really struggle with promoting:

```
const knativeEnvVars = JSON.parse(process.env.KNATIVE_CONTEXT);
const knativeService = knativeEnvVars.K_SERVICE;
const knativeRevision = knativeEnvVars.K_REVISION;
// future
const knativeFooBar = knativeEnvVars.K_FOO_BAR;
```

vs

```
const knativeService = process.env.K_SERVICE;
const knativeRevision = process.env.K_REVISION;
// future addition
const knativeFooBar = process.env.K_FOO_BAR;
```

The JSON parse step is something folks are likely to forget or mess up vs simply accessing the variables separately. System env vars are prefixed appropriately as well (`K_`) so I feel like that's enough of a guide for folks to differentiate system context from their context.

I've never encountered bundled env vars--are there any other products (not ours) that use this pattern?

Reply:
We started this PR to get the conversation going about adding these env variables and implemented it this way initially because It was already suggested in the issue comments and I personally have used this pattern before and didn’t have  issues with it, but we are open to whatever the community prefers.  It is still work in progress as the service name still needs to get the correct value and we can  switch to multiple variables

Reply:
I would love to have multiple ENV vars. Another reason to have multiple vars is: functions will be deployed to Kubernetes cluster. Some vars will be taken from Secrets and some from ConfigMap. It won't be possible to do with only one ENV var.

Reply:
Single env variable does not violate 12 factors apps. Declaring same env vars per environment is not recommended, but this is different from having two fields in a big json variable or in two separate fields

Reply:
Assuming I need to set 2 ENV vars. First comes from Secrets (e.g. database password) https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables , second comes from ConfigMap https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#define-a-container-environment-variable-with-data-from-a-single-configmap or multiple ConfigMaps https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#define-container-environment-variables-with-data-from-multiple-configmaps  . How can I do it with current approach?